File: web/src/workbench/directory-cache.ts
import type { AgentRuntime, Runner } from "../lib/api";
import { targetKey, type AgentTarget } from "./model.ts";

export type DirectoryAgent = { agent_ref?: string; target: AgentTarget; runtime: AgentRuntime; runner?: Runner };
export type DirectoryPage = {
  items: DirectoryAgent[];
  runners: { runner: Pick<Runner, "id" | "binding">; online: boolean; ready: boolean }[];
  issues: { runner_id: string; code: string; runtime?: AgentRuntime }[];
  next_cursor?: string; complete: boolean;
};
export type DirectoryEvent = { subscription_id: string; kind: "ready" | "member" | "metadata" | "invalidated" | "heartbeat"; agent?: DirectoryAgent };
export type DiscoveryBatch = Readonly<{ subscription: string }>;

// A batch is an object capability for one subscription lifetime. Invalidating
// it fences resolved fetches and buffered callbacks, even after reconnect.
export class DirectoryCache {
  private batch?: DiscoveryBatch;
  private members = new Map<string, DirectoryAgent>();
  begin(subscription: string): DiscoveryBatch {
    this.invalidate();
    this.batch = Object.freeze({ subscription });
    return this.batch;
  }
  invalidate() { this.batch = undefined; this.members.clear(); }
  valid(batch: DiscoveryBatch) { return this.batch === batch; }
  mergePage(batch: DiscoveryBatch, page: DirectoryPage): boolean {
    if (!this.valid(batch)) return false;
    for (const agent of page.items) if (!this.merge(agent, true)) return false;
    return true;
  }
  apply(batch: DiscoveryBatch, event: DirectoryEvent): boolean {
    if (!this.valid(batch) || event.subscription_id !== batch.subscription) return false;
    if (event.kind === "invalidated") { this.invalidate(); return true; }
    if (event.kind === "heartbeat") return true;
    return !!event.agent && this.merge(event.agent, event.kind === "member");
  }
  private merge(agent: DirectoryAgent, member: boolean): boolean {
    const { runtime, target } = agent;
    if (!target?.binding?.runner_id || !runtime?.id || runtime.id !== target.runtime.id || runtime.incarnation !== target.runtime.incarnation || runtime.generation !== target.runtime.generation || runtime.adapter !== target.runtime.adapter) return false;
    const key = targetKey(target), previous = this.members.get(key);
    if (!previous && (!member || this.members.size >= 4096)) return false;
    const next = structuredClone(agent), metadata = runtime.session_metadata, old = previous?.runtime.session_metadata;
    if (metadata && !/^(0|[1-9][0-9]*)$/.test(metadata.revision)) return false;
    if (old && metadata && BigInt(metadata.revision) < BigInt(old.revision)) return true;
    if (old && !metadata) {
      next.runtime.session_metadata = structuredClone(old);
      next.agent_ref = previous?.agent_ref;
    }
    this.members.set(key, next);
    return true;
  }
  snapshot(): DirectoryAgent[] { return structuredClone([...this.members.values()]); }
}

File: pkg/agents/directory_cache.go
package agents

import (
	"encoding/json"
	"sort"
	"sync"

	"github.com/aiomni/dune/pkg/workbench"
)

// DiscoveryBatch binds asynchronous pages to one valid subscription lifetime.
// Its fields are private so an old result cannot be relabeled after reconnect.
type DiscoveryBatch struct {
	subscription string
	generation   uint64
}

// DirectoryCache is the SDK's merge boundary. Membership and subscription
// validity are checked before complete observation versions. Returned values
// are owned copies, not mutable references into the cache.
type DirectoryCache struct {
	mu           sync.Mutex
	subscription string
	generation   uint64
	valid        bool
	members      map[workbench.AgentTarget]Agent
}

func (c *DirectoryCache) Begin(subscriptionID string) DiscoveryBatch {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.generation++
	c.subscription, c.valid = subscriptionID, subscriptionID != ""
	c.members = make(map[workbench.AgentTarget]Agent)
	return DiscoveryBatch{subscription: subscriptionID, generation: c.generation}
}

func (c *DirectoryCache) Invalidate(subscriptionID string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.subscription == subscriptionID {
		c.invalidateLocked()
	}
}

func (c *DirectoryCache) invalidateLocked() {
	c.valid = false
	c.generation++
	c.members = nil
}

func (c *DirectoryCache) MergePage(batch DiscoveryBatch, page DirectoryPage) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	if !c.valid || batch.subscription != c.subscription || batch.generation != c.generation {
		return false
	}
	for _, agent := range page.Items {
		if !c.mergeLocked(agent, true) {
			c.invalidateLocked()
			return false
		}
	}
	// Missing items, especially in an incomplete page, are not removals.
	return true
}

func (c *DirectoryCache) Apply(event DirectoryEvent) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	if !c.valid || event.SubscriptionID != c.subscription {
		return false
	}
	if event.Kind == DirectoryInvalidated {
		c.invalidateLocked()
		return true
	}
	if event.Agent == nil || (event.Kind != DirectoryMember && event.Kind != DirectoryMetadata) {
		return false
	}
	if !c.mergeLocked(*event.Agent, event.Kind == DirectoryMember) {
		return false
	}
	return true
}

func (c *DirectoryCache) mergeLocked(next Agent, establishesMembership bool) bool {
	if next.Target.Validate() != nil || next.Runtime.ID != next.Target.Runtime.ID || next.Runtime.Incarnation != next.Target.Runtime.Incarnation || next.Runtime.Generation != next.Target.Runtime.Generation || next.Runtime.Adapter != next.Target.Runtime.Adapter || next.Runner.ID != next.Target.Binding.RunnerID || next.Runner.Binding == nil || *next.Runner.Binding != next.Target.Binding {
		return false
	}
	previous, exists := c.members[next.Target]
	if !exists && (!establishesMembership || len(c.members) >= MaxDirectoryMembers) {
		return false
	}
	if !next.Runtime.Observation.Valid() || (exists && next.Runtime.Observation.Epoch != previous.Runtime.Observation.Epoch) {
		c.invalidateLocked()
		return false
	}
	if exists && next.Runtime.Observation.Revision <= previous.Runtime.Observation.Revision {
		return true
	}
	if exists && previous.Runtime.SessionMetadata != nil && next.Runtime.SessionMetadata == nil {
		// Missing metadata is not an explicit title clear.
		next.Runtime.SessionMetadata = previous.Runtime.SessionMetadata
	}
	c.members[next.Target] = copyAgent(next)
	return true
}

func (c *DirectoryCache) Snapshot() []Agent {
	c.mu.Lock()
	defer c.mu.Unlock()
	items := make([]Agent, 0, len(c.members))
	for _, agent := range c.members {
		items = append(items, copyAgent(agent))
	}
	sort.Slice(items, func(i, j int) bool { return items[i].Target.Key() < items[j].Target.Key() })
	return items
}

func copyAgent(value Agent) Agent {
	data, _ := json.Marshal(value)
	var copied Agent
	_ = json.Unmarshal(data, &copied)
	return copied
}
